borsh-derive-internal/src/enum_ser.rs

```diff
@@ -109,3 +80,196 @@ pub fn enum_ser(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
         }
     })
 }
+
+struct VariantParts {
+    where_predicates: Vec<WherePredicate>,
+    variant_header: TokenStream2,
+    variant_body: TokenStream2,
+    variant_idx_body: TokenStream2,
+}
+fn named_fields(
+    cratename: &Ident,
+    enum_ident: &Ident,
+    variant_ident: &Ident,
+    discriminant_value: &TokenStream2,
+    fields: &FieldsNamed,
+) -> syn::Result<VariantParts> {
+    let mut where_predicates: Vec<WherePredicate> = vec![];
+    let mut variant_header = TokenStream2::new();
+    let mut variant_body = TokenStream2::new();
+    let mut dot_dot_pattern_match: Option<syn::Token![..]> = None;
+    for field in &fields.named {
+        let field_name = if contains_skip(&field.attrs) {
+            dot_dot_pattern_match = Some(syn::Token![..](Span::call_site()));
+            None
+        } else {
+            Some(field.ident.clone().unwrap())
+        };
+        if let Some(ref field_name) = field_name {
+            variant_header.extend(quote! { #field_name, });
+        }
+        if !contains_skip(&field.attrs) {
+            let field_type = &field.ty;
+            where_predicates.push(
+                syn::parse2(quote! {
+                    #field_type: #cratename::ser::BorshSerialize
+                })
+                .unwrap(),
+            );
+            variant_body.extend(quote! {
+                 #cratename::BorshSerialize::serialize(#field_name, writer)?;
+            })
+        }
```

Comment:
This way it is easier to follow the logic.

Note, I also:
* replaced `dot_dot_pattern_match` with `has_skipped_fields` (in fact, I feel we can drop it, and unconditionally add `..` - it will work fine even if all the fields are named explicitly)
* renamed `field_name` to `field_ident` to align with the terminology used in the "unnamed_fields" function)

```suggestion
        if contains_skip(&field.attrs) {
            has_skipped_fields = true;
        } else {
            let field_ident = field.ident.clone().unwrap();
            variant_header.extend(quote! { #field_ident, });
            let field_type = &field.ty;
            where_predicates.push(
                syn::parse2(quote! {
                    #field_type: #cratename::ser::BorshSerialize
                })
                .unwrap(),
            );
            variant_body.extend(quote! {
                 #cratename::BorshSerialize::serialize(#field_ident, writer)?;
            })
        }
```